v0x04/match_fields.py

```diff
@@ -42,7 +42,10 @@ def as_of_tlv(self):
         except ValueError:
             value, mask = map(int, self.value.split('/'))
             oxm_hasmask = True
-        value = value | VlanId.OFPVID_PRESENT
+        if value == 0:
+            value = VlanId.OFPVID_NONE
+        else:
+            value = value | VlanId.OFPVID_PRESENT
         value_bytes = value.to_bytes(2, 'big')
         if mask:
```

Comment:
@Alopalao could you please also take the opportunity to confirm that another mask value other than 4096 also works as expected? Let's say `"2/4094"` -> `0xFFE` -> 0b111111111110, in this case expecting to match vlan 2 or 3

I'm still not sure if the mask should have this logical or, `mask = mask | VlanId.OFPVID_PRESENT` with `4096` it would still be the same so, that's why it's also great to test with other mask values to confirm. In the OF spec in that table that's not immediately clear to me either, but also when comparing with other masked oxm mask values implemented on `of_core` [nw src it's adding the mask int value as bytes as it is](https://github.com/kytos-ng/of_core/blob/master/v0x04/match_fields.py#L194-L196)

This is in the scope of vlan range, but since you're at it, it'll be also helpful for @ajoaoff to know that this part has already been covered/confirmed too.

Reply:
Yes, it works but only when both switches are sending packages with the same vlan (2, 2 or 3, 3). 
For example the code below results in 100% packet loss:
```
py h11.cmd('ip link add link %s name vlan2 type vlan id 2' % (h11.intfNames()[0]))
py h11.cmd('ip link set up vlan2')
py h11.cmd('ip addr add 100.0.0.11/24 dev vlan2')
py h2.cmd('ip link add link %s name vlan3 type vlan id 3' % (h2.intfNames()[0]))
py h2.cmd('ip link set up vlan3')
py h2.cmd('ip addr add 100.0.0.2/24 dev vlan3')
py h11.cmd('ping -c1 100.0.0.2')
```

Reply:
> Yes, it works but only when both switches are sending packages with the same vlan (2, 2 or 3, 3). For example the code below results in 100% packet loss:
> 
> ```
> py h11.cmd('ip link add link %s name vlan2 type vlan id 2' % (h11.intfNames()[0]))
> py h11.cmd('ip link set up vlan2')
> py h11.cmd('ip addr add 100.0.0.11/24 dev vlan2')
> py h2.cmd('ip link add link %s name vlan3 type vlan id 3' % (h2.intfNames()[0]))
> py h2.cmd('ip link set up vlan3')
> py h2.cmd('ip addr add 100.0.0.2/24 dev vlan3')
> py h11.cmd('ping -c1 100.0.0.2')
> ```
> 
> I set up two hosts for each switch with different vlan;
> 
> ```
> h11 - vlan: 3 - s1
> h12 - vlan: 2 - s1
> h21 - vlan: 3 - s2
> h22 - vlan: 2 - s2
> ```
> 
> h11 cannot ping h12 and h22.

Awesome, good to know, that's already great info for @ajoaoff too. Regarding the asymmetric vlans on `any`, that's related to chat we had the other day, right? That Italo also explained that ping isn't the ideal tool here to test that out, however as you pointed out with `ping` and symmetric vlans works. I opened a thread regarding this on https://github.com/kytos-ng/kytos-end-to-end-tests/pull/210, maybe if we add some asserts to UNI flows based on your spreadsheet that be sufficient, I'll ask you to sync with @italovalcy in terms of e2e testing if that's OK. Thanks.

Reply:
> Currently I am only allowing "untagged" and "any". Do I add these type of cases to the `pydantic` implementation?

Yes, @Alopalao, the way you're structuring on your PR looks good. But please, let's make sure it's also evolving aligned with vlan range, it's a separated feature, but considering our mask capabilities that has the potential to reuse the same field cc'ing @ajoaoff for you guys to discuss and propose the way it'll be. 

Also, considering `dl_vlan` being a `str` on `flow_manager` API on PR https://github.com/kytos-ng/flow_manager/pull/135/files,  accepting `value: Union[int, str]` looks good, plus it still keep compatibility with other type of `tag_type` != 1 (without breaking the API, although other tag types aren't configurable on UNI core objects yet), but when it's `tag_type` 1, let's make sure to send this as a `str`.

Reply:
Vlan range is evolving into a more complex requirement allowing more flexibility in the vlan value.